appbuilder/services/livesync/livesync-service.ts

```diff
@@ -17,86 +17,82 @@ export class ProtonLiveSyncService implements IProtonLiveSyncService {
-					errors.push(err);
-				}
-			});
+	private liveSyncOnDevice(deviceDescriptor: IDeviceLiveSyncInfo, filePaths: string[]): IFuture<IDeviceLiveSyncResult> {
+		return ((): IDeviceLiveSyncResult => {
+			let result: IDeviceLiveSyncResult = {
+				deviceIdentifier: deviceDescriptor.deviceIdentifier,
+				liveSyncToApp: null,
+				liveSyncToCompanion: null
+			};
 
-			if(errors && errors.length) {
-				throw new Error(`Unable to livesync to specified devices. Errors are: ${errors.map(err => err.message)}`);
+			let device = _.find(this.$devicesService.getDeviceInstances(), d => d.deviceInfo.identifier === deviceDescriptor.deviceIdentifier);
+			if(!device) {
+				result.liveSyncToApp = result.liveSyncToCompanion = {
+					isResolved: false,
+					error: new Error(`Cannot find connected device with identifier ${deviceDescriptor.deviceIdentifier}.`)
+				};
+
+				return result;
 			}
-		}).future<void>()();
-	}
 
-	private liveSyncToApp(platform: string, devices: Mobile.IDevice[], filePaths: string[]): IFuture<void> {
-		return (() => {
-			this.$options.companion = false;
-			if(!this.$project.capabilities.livesync) {
-				this.$errors.failWithoutHelp(`You cannot use LiveSync for ${this.$project.projectData.Framework} projects.`);
+			let appIdentifier = this.$project.projectData.AppIdentifier,
+				canExecute = (d: Mobile.IDevice) => d.deviceInfo.identifier === device.deviceInfo.identifier,
+				livesyncData: ILiveSyncData = {
+					platform: device.deviceInfo.platform,
+					appIdentifier: appIdentifier,
+					projectFilesPath: this.$project.projectDir,
+					syncWorkingDirectory: this.$project.projectDir,
+					excludedProjectDirsAndFiles: this.excludedProjectDirsAndFiles,
+					canExecuteFastSync: false
+				};
+
+			let canExecuteAction = this.$liveSyncServiceBase.getCanExecuteAction(device.deviceInfo.platform, appIdentifier, canExecute);
+			if(deviceDescriptor.syncToApp) {
+				result.liveSyncToApp = this.liveSyncCore(livesyncData, device, appIdentifier, canExecuteAction, { isForCompanionApp: false }, filePaths).wait();
+			}
+
+			if(deviceDescriptor.syncToCompanion) {
+				result.liveSyncToCompanion = this.liveSyncCore(livesyncData, device, appIdentifier, canExecuteAction, { isForCompanionApp: true }, filePaths).wait();
 			}
 
-			this.liveSyncCore(platform, devices, filePaths).wait();
-		}).future<void>()();
+			return result;
+		}).future<IDeviceLiveSyncResult>()();
 	}
 
-	private liveSyncToCompanion(platform: string, devices: Mobile.IDevice[], filePaths: string[]): IFuture<void> {
+	private liveSyncCore(livesyncData: ILiveSyncData, device: Mobile.IDevice, appIdentifier: string, canExecuteAction: (dev: Mobile.IDevice) => boolean, liveSyncOptions: { isForCompanionApp: boolean }, filePaths: string[]): IFuture<ILiveSyncOperationResult> {
 		return (() => {
-			this.$options.companion = true;
+			let liveSyncOperationResult = <ILiveSyncOperationResult>{};
```

Comment:
what @teobugslayer suggests cannot compile, I believe, but you could go with
````TypeScript
let liveSyncOperationResult: ILiveSyncOperationResult = { isResolved: false }
```
and go ahead with the logic below whilst skipping any unnecessary reassignment to `false`